### PR TITLE
Fix sbang for perl

### PR DESCRIPTION
**Pull request by muffgaga**
We encountered issues with spack's `git` installation when installing into deep directory hierarchies.
(Example: `spack install git && spack load git && git add -p` fails with `bad interpreter` error)

Two issues were found:
- git helpers (located under `libexec/git-core/*`) were not sbang-adjusted but only `bin/*`
- the perl interpreter needs an additional flag (`-x`, cf. perlrun docs)

### CODE REVIEW

**Comment by muffgaga** [comment]
Hmm, we filtered out non-executable files => we'll adapt the change to the current upstream behavior.

**Comment by tgamblin** [comment]
@muffgaga: looks like this breaks the `sbang` tests -- run `spack test sbang` (or look at the Travis output).

**Comment by muffgaga** [comment]
@tgamblin I rebased the patch set and removed the filter on executables. Ok? :smile:

File: lib/spack/spack/hooks/sbang.py [hunk 1]
```diff
@@ -81,8 +81,9 @@ def filter_shebang(path):
     tty.warn("Patched overlong shebang in %s" % path)
 
 
-def filter_shebangs_in_directory(directory):
-    for file in os.listdir(directory):
+def filter_shebangs_in_directory(directory, filenames=None):
+    filenames = filenames or os.listdir(directory)
```
Comment (tgamblin) [inline lib/spack/spack/hooks/sbang.py]: This probably results in unexpected behavior if the caller programmatically constructs a list and it ends up being an empty list.  Can you change this to only list if filenames is `None`?

**Comment by muffgaga** [comment]
ACK, good point. Fixed.

**Pull request merged**